### ISSUE TITLE
`Concatenate` false positive (no problem with `pyright`)

### ISSUE THREAD

**Issue by supersergiy** [python/mypy#14168]
**Bug Report**

`Concatenate` causes a false positive in the following example:

**To Reproduce**
[Playground Link](https://mypy-play.net/?mypy=latest&python=3.11&gist=f68db157037332084da52bb872c54149)
```python
from typing import Protocol
from typing_extensions import ParamSpec, Concatenate

P = ParamSpec("P")

class A(Protocol[P]):
    def foo(self, a: int, *args: P.args, **kwargs: P.kwargs):
        ...
        
class B(Protocol[P]):
    def foo(self, a: int, b: int, *args: P.args, **kwargs: P.kwargs):
        ...
    
def bar(b: B[P]) -> A[Concatenate[int, P]]:
    return b # error
```

**Expected Behavior**

No type error expected, which works with `pyright` ([pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAKeEfWMC_6VQ2wrCMAx971cEn7ox-wEDhc0fKPg4hnS1HcOtGW1F_Xtbd_NRMBBykpOcA9EWB_CvsTMtdMOI1gO36FFiT_TGXdTTK-M6NG5dE1YM51HJDE5opAh8SEI4HDaO7vguIUT2wjko6CJd8TrJCYS4Kg0akTrV6wxEDp3xGaTCti4HzmINbXp7LJMJzdcxGGMrno3Kn42aPwxJlGyEpUGkjEawP0JRfT2j-mjzup6OrfJ3a6B5A0JbxGd0AQAA))

**Actual Behavior**

```
# main.py:15: error: Incompatible return value type (got "B[P]", expected "A[[int, **P]]")  [return-value]
# main.py:15: note: Following member(s) of "B[P]" have conflicts:
# main.py:15: note:     Expected:
# main.py:15: note:         def foo(self, a: int, *args: [int, **P.args], **kwargs: [int, **P.kwargs]) -> Any
# main.py:15: note:     Got:
# main.py:15: note:         def foo(self, a: int, b: int, *args: P.args, **kwargs: P.kwargs) -> Any
# Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**
- Mypy version used: 0.991

**Comment by ilevkivskyi** [comment]
On master mypy emits a better error now:
```
error: Incompatible return value type (got "B[P]", expected "A[[int, **P]]")  [return-value]
note: Following member(s) of "B[P]" have conflicts:
note:     Expected:
note:         def foo(a: int, int, /, *args: P.args, **kwargs: P.kwargs) -> Any
note:     Got:
note:         def foo(self, a: int, b: int, *args: P.args, **kwargs: P.kwargs) -> Any
```
I think this should be allowed, function with named argument should be fine where a function with positional-only argument is expected (unless I am missing something). And this error also hints a workaround: making `a` and `b` arguments in both `foo` positional-only (e.g. using `__a` and `__b` argument names) makes the error go away.

**Comment by ilevkivskyi** [comment]
OK, after reading the code, it looks like there is some unsafety here. Consider for example `A[Concatenate[int, ...]]` (btw I think we don't support this syntax yet) or similar, then the signature of `A.foo` is
```python
def foo(self, a: int, int, *args: Any, **kwargs: Any)
```
and `foo(b="what?")` is a valid call for this signature (because second argument is positional-only, so `b` will map to `**kwargs`), while it is obviously not valid for `B`. This is a super rare corner-case however, and funny part is that this check should be only enabled in `--strict` mode, but it was accidentally always on in few places.

**Comment by hauntsaninja** [comment]
@supersergiy is the project you were running into this on open source?

**Comment by supersergiy** [comment]
@hauntsaninja yup it is, why?

**Comment by hauntsaninja** [comment]
I'd like to add it to https://github.com/hauntsaninja/mypy_primer . This will let us detect regressions or measure future improvements on your code. I think we could use more coverage of real world ParamSpec usage.

**Comment by supersergiy** [comment]
Yup, that's the one. Happy if it helps

**Comment by hauntsaninja** [comment]
It has already helped :-) https://github.com/python/mypy/pull/15953#issuecomment-1692463886
